packages/sanity/src/core/store/bundles/useBundleOperations.ts

```diff
@@ -59,19 +64,78 @@ export function useBundleOperations() {
         .filter(([_, value]) => value === undefined)
         .map(([key]) => key)
 
-      let clientOperation = client.patch(bundle._id).set(document)
+      let clientOperation = addOnClient.patch(bundle._id).set(document)
       if (unsetKeys.length) {
         clientOperation = clientOperation.unset(unsetKeys)
       }
 
       return clientOperation.commit()
     },
-    [client],
+    [addOnClient],
+  )
+
+  const handlePublishBundle = useCallback(
+    async (
+      bundleId: string,
+      bundleDocuments: SanityDocument[],
+      publishedDocumentsRevisions: Record<string, string> = {},
+    ) => {
+      if (!addOnClient) return null
+
+      const transaction = studioClient.transaction()
+      bundleDocuments.forEach((bundleDocument) => {
+        const publishedDocumentId = getPublishedId(bundleDocument._id, true)
+        const versionDocument = omit(bundleDocument, ['_version']) as SanityDocument
+        const publishedDocumentRevisionId = publishedDocumentsRevisions[publishedDocumentId]
+
+        const publishedDocument = {
+          ...versionDocument,
+          _id: publishedDocumentId,
+        }
+        // verify that local bundle document matches remote latest revision
+        transaction.patch(bundleDocument._id, {
+          unset: ['_revision_lock_pseudo_field_'],
+          ifRevisionID: bundleDocument._rev,
+        })
+
+        if (publishedDocumentRevisionId) {
+          // if published document exists, verify that local document matches remote latest revision
+          transaction.patch(publishedDocumentId, {
+            unset: ['_revision_lock_pseudo_field_'],
+            ifRevisionID: publishedDocumentRevisionId,
+          })
+          // update the published document with the bundle version
+          transaction.createOrReplace(publishedDocument)
+        } else {
+          // if published document doesn't exist, do not override
+          // only create the document and fail is it suddenly exists
+          transaction.create(publishedDocument)
+        }
+      })
+
+      await transaction.commit()
+      const publishedAt = new Date().toISOString()
+      return await addOnClient.patch(bundleId).set({publishedAt, archivedAt: publishedAt}).commit()
+    },
+    [addOnClient, studioClient],
   )
 
+  const guardForAddOnClient = <Callback extends (...args: any[]) => any>(
```

Comment:
Does this needs to be an useCallback?

Reply:
Yeh, good callout! I've changed the implementation here since trying to memoise the return from the HOF was proving a little unreadable and clunky